modules/cockroachdb/src/main/java/org/testcontainers/containers/CockroachContainer.java

```diff
@@ -104,22 +107,57 @@ public String getTestQueryString() {
 
     @Override
     public CockroachContainer withUsername(String username) {
-        throw new UnsupportedOperationException(
-            "The CockroachDB docker image does not currently support this - please see https://github.com/cockroachdb/cockroach/issues/19826"
-        );
+        validateIfVersionSupportsUsernameOrPasswordOrDatabase(getDockerImageName(), "username");
+        this.username = username;
+        return withEnv("COCKROACH_USER", username);
     }
 
     @Override
     public CockroachContainer withPassword(String password) {
-        throw new UnsupportedOperationException(
-            "The CockroachDB docker image does not currently support this - please see https://github.com/cockroachdb/cockroach/issues/19826"
-        );
+        validateIfVersionSupportsUsernameOrPasswordOrDatabase(getDockerImageName(), "password");
+        this.password = password;
+        return withEnv("COCKROACH_PASSWORD", password).withCommand("start-single-node");
     }
 
     @Override
     public CockroachContainer withDatabaseName(final String databaseName) {
-        throw new UnsupportedOperationException(
-            "The CockroachDB docker image does not currently support this - please see https://github.com/cockroachdb/cockroach/issues/19826"
-        );
+        validateIfVersionSupportsUsernameOrPasswordOrDatabase(getDockerImageName(), "databaseName");
+        this.databaseName = databaseName;
+        return withEnv("COCKROACH_DATABASE", databaseName);
+    }
+
+    private void validateIfVersionSupportsUsernameOrPasswordOrDatabase(String dockerImageName, String parameter) {
```

Comment:
I think in constructor at line 53, `ComparableVersion` can be used and store the result in `isVersionGreaterThanOrEqualTo221` and use it in `withUsername`, `withPassword` and `withDatabaseName` in order to throw the `UnsupportedOperationException `. In order to get the version, call `dockerImageName.getVersionPart()` and drop the `v` from it.

Reply:
Thanks, that's indeed an improvement! Applied.